### PR TITLE
Emit .loc directives for out of line code

### PR DESCRIPTION
**Pull request by gretay-js**
Emit correct location information for cold code part of `Alloc` and `Checkbound`.  

The code emitted for `Alloc` and `Checkbound` includes a condition and a call to the approprate one of the following
- `caml_call_gc`
- `caml_call_local_realloc`
- `caml_ml_array_bound_error`
These calls are cold and emitted in as  separate blocks out of line at the end of the function. These blocks does not have `.loc` directives, so they gets associated with source location information of whatever the last `.loc` directive at the end of the function was. This messes up FDO counters. 

This PR adds .loc directives for them with the same source information as their original Linear instruction. 
FDO to distinguish the orignal instruction from the curresponding code block, the discriminator of the cold block is negative. FDO optimizations that interpret these discriminators live elsewhere and will be updated accordingly.

### CODE REVIEW

File: backend/amd64/emit.mlp [hunk 1]
```diff
@@ -65,10 +65,13 @@ let cfi_adjust_cfa_offset n =
 let emit_debug_info ?discriminator dbg =
   emit_debug_info_gen ?discriminator dbg D.file D.loc
 
-let emit_debug_info_linear i =
-  match i.fdo with
-  | None -> emit_debug_info i.dbg
+let emit_debug_info_instr ~pad fdo dbg =
```
Comment (poechsel) [inline backend/amd64/emit.mlp:68]: I don't quite like `~pad` as it's pretty non-descriptive. Would it instead be possible to have a second function,
`emit_debug_info_synthetic_block` what will always write the negative discriminator and keep `emit_debug_info_linear` as-is?

Comment (gretay-js) [reply backend/amd64/emit.mlp:68]: agree that `pad` is a bad name, I'll try to find a better one and refactor this little function. I don't want to duplicate the logic that decides which location to emit.

File: backend/amd64/emit.mlp [hunk 2]
```diff
@@ -65,12 +65,18 @@ let cfi_adjust_cfa_offset n =
...
 
-let emit_debug_info_linear i =
-  match i.fdo with
-  | None -> emit_debug_info i.dbg
+let emit_debug_info_helper ~out_of_line fdo dbg =
+  match (fdo : Fdo_info.t) with
+  | None -> emit_debug_info dbg
   | Some { discriminator; dbg } ->
+    let discriminator =
+      if out_of_line then (-discriminator) else  discriminator
```
Comment (xclerc) [inline backend/amd64/emit.mlp:73]:
```suggestion
      if out_of_line then (-discriminator) else discriminator
```
(Should we enable `ocamlformat` on that file,
or is it disabled to keep merges manageable?)

**Review by xclerc — APPROVED** [review]

File: backend/amd64/emit.mlp (hunk 2 above)
Comment (gretay-js) [reply backend/amd64/emit.mlp]: yes, I want to enable on all of `backend` folder  recursively.

**Pull request merged**